### PR TITLE
os/bluestore: fix collection_list ordering

### PR DESCRIPTION
**Pull request by trociny**
Fixes: https://tracker.ceph.com/issues/43174
Signed-off-by: Mykola Golub <mgolub@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by ifed01** [comment]
jenkins test make check

**Comment by ifed01** [comment]
This doesn't pass 'store_test' suite, ObjectStore/StoreTest.FiemapHoles/2 test case specifically:

/home/if/ceph/src/os/bluestore/BlueStore.cc: 719: FAILED ceph_assert(r == 0)
 ceph version Development (no_version) pacific (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x19d) [0x7f62ba7bf1e1]
 2: (()+0x170a463) [0x7f62ba7bf463]
 3: (()+0xe22e46) [0x56428568ee46]
 4: (()+0xe22975) [0x56428568e975]
 5: (()+0xe22afd) [0x56428568eafd]
 6: (BlueStore::_collection_list(BlueStore::Collection*, ghobject_t const&, ghobject_t const&, int, std::vector<ghobject_t, std::allocator<ghobject_t> >*, ghobject_t*)+0xef9) [0x5642856f4ed7]
 7: (BlueStore::_remove_collection(BlueStore::TransContext*, coll_t const&, boost::intrusive_ptr<BlueStore::Collection>*)+0x30e) [0x56428572b5c4]
 8: (BlueStore::_txc_add_transaction(BlueStore::TransContext*, ceph::os::Transaction*)+0x3bc) [0x56428570d456]
 9: (BlueStore::queue_transactions(boost::intrusive_ptr<ObjectStore::CollectionImpl>&, std::vector<ceph::os::Transaction, std::allocator<ceph::os::Transaction> >&, boost::intrusive_ptr<TrackedOp>, ThreadPool::TPHandle*)+0x358) [0x56428570c588]
 10: (ObjectStore::queue_transaction(boost::intrusive_ptr<ObjectStore::CollectionImpl>&, ceph::os::Transaction&&, boost::intrusive_ptr<TrackedOp>, ThreadPool::TPHandle*)+0x93) [0x56428550f0ed]
 11: (int queue_transaction<boost::scoped_ptr<ObjectStore> >(boost::scoped_ptr<ObjectStore>&, boost::intrusive_ptr<ObjectStore::CollectionImpl>, ceph::os::Transaction&&)+0xc5) [0x56428552c50f]
 12: (StoreTest_FiemapHoles_Test::TestBody()+0x139d) [0x564285459afd]
...
 ../../store_test.sh: line 21:  6599 Aborted                 (core dumped) bin/ceph_test_objectstore --gtest_filter=ObjectStore/StoreTest.FiemapHoles/2 --gtest_break_on_failure

**Comment by ifed01** [comment]
I'd highly appreciate having some test case coverage for the sorting issue and the fix...

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -596,6 +617,121 @@ ostream& operator<<(ostream& out, const BlueStore::Buffer& b)
...
+    ghobject_t oid;
+    int r = get_key_object(key, &oid);
+    ceph_assert(r != -1);
+
+    std::string pkey;
+    _key_encode_prefix(oid, &pkey);
+
+    m_it->lower_bound(pkey);
+    if (!get_next_chunk()) {
+      return;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:658]: Perhaps you need to "invalidate" (i.e. set m_key to nullptr) at this point

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -596,6 +617,121 @@ ostream& operator<<(ostream& out, const BlueStore::Buffer& b)
...
+    }
+
+    auto iter = m_chunk.find(oid);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:661]: shouldn't this be lower_bound not find? to proceed with the next entry after oid if the latter is missing.

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -596,6 +617,121 @@ ostream& operator<<(ostream& out, const BlueStore::Buffer& b)
...
+    lower_bound(key);
+
+    if (valid() && *m_key == key) {
+      next();
+    }
+  }
+
+  void next() {
+    if (m_chunk_iter == m_chunk.end()) {
+      if (!get_next_chunk()) {
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:681]: isn't get_next_chunk() call redundant? get_next_chunk() func looks like a single-shot stuff. It wouldn't retrieve additional data after the first call unless KV's iterator is reset by lower_bound. Hence no need for its call from next() func.

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -596,6 +617,121 @@ ostream& operator<<(ostream& out, const BlueStore::Buffer& b)
...
+ * the KeyValueDB iterator does not lexicographically sort the same
+ * way that ghobject_t does: objects with the same hash may have wrong order.
+ *
+ * This is the iterator wrapper that fixes the keys order.
+ */
+
+class CollectionListIterator {
+public:
+  CollectionListIterator(const KeyValueDB::Iterator &it)
+    : m_it(it), m_chunk_iter(m_chunk.begin()) {
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:633]: nit: using m_chunk.end() would look clearer to me

File: src/os/bluestore/BlueStore.cc (hunk 3 above)
Comment (trociny) [reply src/os/bluestore/BlueStore.cc:681]: I am not sure I understand your point. A user (bluestore) may call `upper/lower_bound()` and reset the KV iterator), and may just call `next()` and then if the chunk is fully processed we need to load the next chunk (advancing the KV iterator, which `get_next_chunk()` does, see `m_it->next()` there).

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (trociny) [reply src/os/bluestore/BlueStore.cc:658]: indeed. Thanks. It looks like better to do this invalidation in `get_next_chunk()` itself.

File: src/os/bluestore/BlueStore.cc (hunk 2 above)
Comment (trociny) [reply src/os/bluestore/BlueStore.cc:661]: We have two possible cases:

1) The `key` exists in the db. In this case, after `m_it->lower_bound(pkey)` above, we load a sorted chunk of key with the same hash as the `key` has (and the `key` is among them).  And `find` will set the iterator to the `key` (as expected).

2) The `key` does not exist in the db. In this case, after `m_it->lower_bound(pkey)` above, we load a sorted chunk of key with the hash that is the next after the `key`'s hash. And in this case we need to have the iterator to set to the first element of this chunk.

File: src/os/bluestore/BlueStore.cc (hunk 3 above)
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:681]: never mind, I missed that  loop get_next_chunk's loop terminates when hash value changes...

**Comment by trociny** [comment]
I update the patch. Now it passes `ceph_test_objectstore --gtest_filter='*/2'`.

The plan is to add tests to cover the sorting issue. The problem is that I may not use object name examples (the same hash and wrong order after encoding)  from the customer and need to generate some new ones myself.

File: src/os/bluestore/BlueStore.cc (hunk 2 above)
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:666]: what's about the case when there is no exact key match but there are multiple different oids with the same hash as for the original input key. Hence you might need to find proper chunk, i.e. chunk next(!) to the input one not the first one

Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:666]: hence you need to replace find with lower_bound AND proceed with the next chunk if this lower_bound returns end() iterator

File: src/os/bluestore/BlueStore.cc (hunk 3 above)
Comment (trociny) [reply src/os/bluestore/BlueStore.cc]: Indeed. Thanks. Updated.

File: src/os/bluestore/BlueStore.cc (hunk 2 above)
Comment (trociny) [reply src/os/bluestore/BlueStore.cc]: Indeed. Thanks. Updated.

**Comment by trociny** [comment]
I have added  the test to cover the sorting issue. Note, it is disabled for kstore, because it is affected the same way (and not fixed yet).

@tchaikov please use the updated branch for testing

**Comment by tchaikov** [comment]
@trociny ack.

**Comment by trociny** [comment]
Couldn't the crash be related to #35659 you are also testing? I don't see from the backtrace where exactly in  `BlueStore::_do_truncate` fails but that PR has many changes for `BlueStore::_do_truncate`.

**Comment by ifed01** [comment]
I bet it's  rather  #35659 indeed

**Comment by tchaikov** [comment]
@ifed01 thanks for looking into the failure. indeed,  #35659  is very suspicious. i dropped it instead.

**Comment by jdurgin** [comment]
To make this work with bluestore upgrades (where old bluestore uses the old order, and osds with this fix don't want to introduce different ordering until all osds support it) do we need to add a feature bit like discussed here to keep the ordering consistent across osds: https://github.com/ceph/ceph/pull/35938#issuecomment-656183558

**Comment by trociny** [comment]
@jdurgin I don't have a strong opinion about this. If it had been only for deep-scrub I would have not bother with it. But as I understand it may also affect the recovery?

Sure I can work on this if people think we need this. There is a technical question though, which I have already risen in the https://github.com/ceph/ceph/pull/35938#issuecomment-657397900: how to pass the feature bit to the bluestore layer in not very ugly way. I would like to have your advise here.

Alternatively we could have it just as a config option (similarly how I did in that PR), disabled by default but with a release note that people may enable it during upgrade if they fill they may be affected.

Also may be there are other ways than a feature bit? I suppose on a pg peering stage the osds could learn if all osds in the set support the new sorting method and would use it then. It might be more complicated (I am not very familiar with this layer) but it would no require to explicitly enable the new sorting via 'ceph osd set ' command -- much nicer for users.

**Comment by jdurgin** [comment]
@trociny due to the potential for missing objects from backfill, I think it's needed.

In terms of passing that information to the objectstore, I don't see a great solution either. Adding an argument to collection_list seems like the least bad approach.

A config option to change the listing order would be possible too, though I'd be concerned about races or misconfigs where some osds applied it at different times. The nice thing about the feature bit approach is we can prevent old osds without the support from booting at all.

Thinking it through now, I like your last suggestion best - since it avoids an extra step for users. Because we're not changing persistent structures, and the uses of collection_list are inherently per-pg, going by the features supported by the acting_backfill_recovery set should be fine. OSDMap has support for caching the up_osd_features - I'm not sure we'd need to cache in this case, but we'd need another method there to include the backfill/async recovery osds too.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-07-26_11:13:42-rados-wip-kefu-testing-2020-07-26-1757-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-07-27_03:46:29-rados-wip-kefu-testing-2020-07-27-0118-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-07-27_15:50:48-rados-wip-kefu-testing-2020-07-27-2127-distro-basic-smithi/

failure tracked by

- https://tracker.ceph.com/issues/46405

**Comment by trociny** [comment]
@jdurgin I will try the last suggestion. Though it might will take some time as I am not very familiar with this field.

**Comment by tchaikov** [comment]
/a/kchai-2020-07-29_04:08:59-rados-wip-kefu-testing-2020-07-29-1027-distro-basic-smithi/5266033/remote/*/log/valgrind/osd.1.log.gz

```xml
<error>
  <unique>0x0</unique>
  <tid>54</tid>
  <threadname>tp_osd_tp</threadname>
  <kind>UninitCondition</kind>
  <what>Conditional jump or move depends on uninitialised value(s)</what>
  <stack>
    <frame>
      <ip>0xBE53A7</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>decode_escaped(char const*, std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt;*)</fn>
      <dir>/usr/src/debug/ceph-16.0.0-3918.g5ad7a922521.el8.x86_64/src/os/bluestore</dir>
      <file>BlueStore.cc</file>
      <line>235</line>
    </frame>
    <frame>
      <ip>0xBE661F</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>int get_key_object&lt;std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; &gt;(std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; const&amp;, ghobject_t*)</fn>
      <dir>/usr/src/debug/ceph-16.0.0-3918.g5ad7a922521.el8.x86_64/src/os/bluestore</dir>
      <file>BlueStore.cc</file>
      <line>375</line>
    </frame>
    <frame>
      <ip>0xBE70C6</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>(anonymous namespace)::CollectionListIterator::lower_bound(std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; const&amp;)</fn>
      <dir>/usr/src/debug/ceph-16.0.0-3918.g5ad7a922521.el8.x86_64/src/os/bluestore</dir>
      <file>BlueStore.cc</file>
      <line>654</line>
    </frame>
    <frame>
      <ip>0xC20A2A</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>upper_bound</fn>
      <dir>/usr/src/debug/ceph-16.0.0-3918.g5ad7a922521.el8.x86_64/src/os/bluestore</dir>
      <file>BlueStore.cc</file>
      <line>679</line>
    </frame>
    <frame>
      <ip>0xC20A2A</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>BlueStore::_collection_list(BlueStore::Collection*, ghobject_t const&amp;, ghobject_t const&amp;, int, std::vector&lt;ghobject_t, std::allocator&lt;ghobject_t&gt; &gt;*
, ghobject_t*)</fn>
      <dir>/usr/src/debug/ceph-16.0.0-3918.g5ad7a922521.el8.x86_64/src/os/bluestore</dir>
      <file>BlueStore.cc</file>
      <line>10716</line>
    </frame>
    <frame>
      <ip>0xC22927</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>BlueStore::collection_list(boost::intrusive_ptr&lt;ObjectStore::CollectionImpl&gt;&amp;, ghobject_t const&amp;, ghobject_t const&amp;, int, std::vector&lt;ghobject_t, s
td::allocator&lt;ghobject_t&gt; &gt;*, ghobject_t*)</fn>
      <dir>/usr/src/debug/ceph-16.0.0-3918.g5ad7a922521.el8.x86_64/src/os/bluestore</dir>
      <file>BlueStore.cc</file>
      <line>10670</line>
    </frame>
    <frame>
      <ip>0x925C98</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>PGBackend::objects_list_range(hobject_t const&amp;, hobject_t const&amp;, std::vector&lt;hobject_t, std::allocator&lt;hobject_t&gt; &gt;*, std::vector&lt;ghobject_t, st
d::allocator&lt;ghobject_t&gt; &gt;*)</fn>
      <dir>/usr/src/debug/ceph-16.0.0-3918.g5ad7a922521.el8.x86_64/src/osd</dir>
      <file>PGBackend.cc</file>
      <line>404</line>
    </frame>
...
```

**Comment by trociny** [comment]
@tchaikov Thanks. It turned out `get_key_object` was not able to detect the end for temp keys properly. I suppose previously it was not used for temp keys thus the problem was not detected. I pushed the fix.

**Comment by trociny** [comment]
@jdurgin Could you please look at my last commit, if it makes sense to you?

Right now it uses `PGBackend::min_peer_features()` to check if all osds have the feature. I plan to implement `min_acting_recovery_backfill_features()` in a similar way and use instead, if this looks good to you.

**Comment by jdurgin** [comment]
Yup, the last commit / approach looks good to me.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-07-30_15:58:21-rados-wip-kefu-testing-2020-07-30-2107-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-07-31_01:42:48-rados-wip-kefu-testing-2020-07-30-2107-distro-basic-smithi/

failure tracked by

- https://tracker.ceph.com/issues/46323

**Comment by trociny** [comment]
We found an issue with my patch when testing it with real data. I pushed the fix for it ("fix end comparison"). It is as a separate commit for now just to make the review easier.

Also, when testing my fix I found an issue with the `_collection_list` that it always was setting `next` to `GMAX` if `max` was not reached (usually when you have `end` set to not `GMAX` and `max` set to `INT_MAX`), which I think is a bug (at least the filestore does not show this behavior). I pushed a patch for this too.

And now, after all my changes, `_collection_list` starts looking a little ugly to me. I think I will work on refactoring it -- get rid of using keys in `_collection_list`, so it would operate with oids only and pass the oids to the iterator, which would do all necessary conversion to keys. I am going on vacation the next two weeks though and I am not sure I will have it ready by then.

**Comment by jdurgin** [comment]
Thinking about the features more, I believe using the existing upacting_features will be sufficient -- any async recovery or backfill target will be a member of the up set. @neha-ojha can you verify this?

This is good, because the acting_recovery_backfill set is not available on replicas; choose_acting() is only called by the primary.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-08-02_10:33:38-rados-wip-kefu-testing-2020-08-02-1650-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-08-03_03:37:05-rados-wip-kefu-testing-2020-08-03-0931-distro-basic-smithi/

**Comment by neha-ojha** [comment]
> Thinking about the features more, I believe using the existing upacting_features will be sufficient -- any async recovery or backfill target will be a member of the up set. @neha-ojha can you verify this?

That's correct. 

> 
> This is good, because the acting_recovery_backfill set is not available on replicas; choose_acting() is only called by the primary.

**Comment by tchaikov** [comment]
@trociny ping? shall we switch over to `upacting_features` in this changeset as suggested by Josh and Neha?

**Comment by trociny** [comment]
@tchaikov I just returned from my vacations and going to work on it this week.

**Comment by trociny** [comment]
> i suspect that this change failed the test of LibRadosList.ListObjectsCursor

Thanks. Indeed. It was caused by my change to the comparison operator for `ghobject_t` to treat `NO_GEN` generation as the smallest one. By some reason `rados_nobjects_list_next` calls the `collection_list` with the `start` object having `generation` set to 0. As a result after my change the `collection_list` skipped the existing object that had all fields as the `start` but  `generation` set to `NO_GEN`.

I removed my change for `ghobject_t`'s `cmp` function. Instead I changed the bluestore's `get_coll_range` to set `generation` to 0 for `start/end` objects (basically, what `get_coll_key_range` previously did) -- it has fixed my initial problem with the assertion failure.

**Comment by tchaikov** [comment]
```
2020-08-26T07:34:26.974 INFO:tasks.workunit.client.0.smithi098.stderr://home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test_kvstore_tool.sh:65: test_ceph_kvstore_tool:  ceph-kvstore-tool bluest
ore-kv ./cephtool.j5Y list
2020-08-26T07:34:26.974 INFO:tasks.workunit.client.0.smithi098.stderr://home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test_kvstore_tool.sh:65: test_ceph_kvstore_tool:  wc -l
2020-08-26T07:34:27.045 INFO:tasks.workunit.client.0.smithi098.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test_kvstore_tool.sh:65: test_ceph_kvstore_tool:  current_kv_nums=0
2020-08-26T07:34:27.045 INFO:tasks.workunit.client.0.smithi098.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test_kvstore_tool.sh:66: test_ceph_kvstore_tool:  test 12 -eq 0
2020-08-26T07:34:27.046 INFO:tasks.workunit.client.0.smithi098.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test_kvstore_tool.sh:1: test_ceph_kvstore_tool:  rm -fr ./cephtool.j5Y
2020-08-26T07:34:27.047 DEBUG:teuthology.orchestra.run:got remote process result: 1
```

/a//kchai-2020-08-26_05:51:42-rados-wip-kefu-testing-2020-08-26-1042-distro-basic-smithi/5376840

**Comment by trociny** [comment]
@tchaikov I am not sure it is related. The `ceph-kvstore-tool` seems to work with the db directly, while my changes are a layer above. I don't see where the bluestore `collection_list` could be used in this test.

BTW, I tried this test locally and it succeeded.

@ifed01 Do you have an idea about this failure?

**Comment by ifed01** [comment]
> @tchaikov I am not sure it is related. The `ceph-kvstore-tool` seems to work with the db directly, while my changes are a layer above. I don't see where the bluestore `collection_list` could be used in this test.
> 
> BTW, I tried this test locally and it succeeded.
> 
> @ifed01 Do you have an idea about this failure?


@tchaikov - wondering if any other PRs were included in this run?

**Comment by tchaikov** [comment]
@ifed01 could be. there are some other PR also could fail this very test. see https://github.com/ceph/ceph/labels/wip-kefu-testing.  i will rerun it without other PR.

**Comment by tchaikov** [comment]
@ifed01 could be. there are some other PR also could fail this very test. see https://github.com/ceph/ceph/labels/wip-kefu-testing.  i will rerun it without other PR.

**Comment by ifed01** [comment]
> @ifed01 could be. there are some other PR also could fail this very test. see https://github.com/ceph/ceph/labels/wip-kefu-testing. i will rerun it without other PR.

Yesterday I made some kvstore-tool related fixes for https://github.com/ceph/ceph/pull/36745. The tool had been failing before that - see @batrick's comment there. So maybe you've been using the previous version of my PR?

**Comment by tchaikov** [comment]
> Yesterday I made some kvstore-tool related fixes for #36745. The tool had been failing before that - see @batrick's comment there. So maybe you've been using the previous version of my PR?

```console
$ git br --contains 30fbd10
  wip-kefu-testing-2020-08-26-1042
  wip-kefu-testing-2020-08-26-1944
  wip-kefu-testing-2020-08-27-0139
```

and test_kvstore_tool.sh still failed in https://pulpito.ceph.com/kchai-2020-08-27_03:37:15-rados-wip-kefu-testing-2020-08-27-0139-distro-basic-smithi/

confirmed this change is not the offending one. i suspect it's #36745. will rerun the tests without #36745

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-08-25_04:54:49-rados-wip-kefu-testing-2020-08-25-0950-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-08-26_16:53:29-rados-wip-kefu-testing-2020-08-26-1944-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-08-27_03:37:15-rados-wip-kefu-testing-2020-08-27-0139-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-08-27_08:22:12-rados-wip-kefu-testing-2020-08-27-1255-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-08-27_10:01:54-rados-wip-kefu-testing-2020-08-27-1255-distro-basic-smithi/

**Comment by tchaikov** [comment]
@trociny regarding "DNM", what is blocking? what are you pending on?

**Comment by trociny** [comment]
@tchaikov I just forgot to remove DNM. Thanks.

Still, it would be nice to see more review.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-08-28_06:18:09-rados-wip-kefu-testing-2020-08-28-1247-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-08-28_13:09:24-rados-wip-kefu-testing-2020-08-28-1247-distro-basic-smithi/

**Comment by trociny** [comment]
> * https://pulpito.ceph.com/kchai-2020-08-28_06:18:09-rados-wip-kefu-testing-2020-08-28-1247-distro-basic-smithi/
> * https://pulpito.ceph.com/kchai-2020-08-28_13:09:24-rados-wip-kefu-testing-2020-08-28-1247-distro-basic-smithi/

The errors like
```
"2020-08-28T07:02:13.843696+0000 osd.9 (osd.9) 38 : cluster [ERR] 3.cs0 shard 4(4) 3:3f45dd62:::smithi04717497-21:177 : missing" in cluster log
```
look related. I am investigating. Thanks.

**Comment by trociny** [comment]
I fixed the "missing" errors. It was a problem with my `SortedCollectionListIterator`. I had a wrong idea about extent shard keys and assumed they wouldn't mix with oid keys for the same hash. As a result an extent shard key could split a chunk providing wrong results. I fixed it and reran the failed tests:

https://pulpito.ceph.com/trociny-2020-08-29_19:23:12-rados-wip-mgolub-testing-4-distro-basic-smithi/
https://pulpito.ceph.com/trociny-2020-08-29_19:23:54-rados-wip-mgolub-testing-4-distro-basic-smithi/

There are no "missing" errors anymore. Still, there are 2 type failures:

1) "reached maximum tries (180) after waiting for 180 seconds"
2) "racing read got wrong version" crash in ceph_test_rados

which I am not sure are related.

**Comment by tchaikov** [comment]
@trociny thanks for looking into the failures. these two failures you listed are known ones.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-09-03_14:29:24-rados-wip-kefu-testing-2020-09-02-1016-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-09-04_15:27:05-rados-wip-kefu-testing-2020-09-04-1852-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-09-05_14:56:36-rados-wip-kefu-testing-2020-09-05-2016-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-09-06_09:34:33-rados-wip-kefu-testing-2020-09-06-1546-distro-basic-smithi/

failures tracked by

- https://tracker.ceph.com/issues/47180
- https://tracker.ceph.com/issues/47179

**Review by jdurgin — APPROVED** [review]

**Comment by tchaikov** [comment]
@jdurgin @ifed01 shall we backport this change?

**Comment by jdurgin** [comment]
@tchaikov yes, updated the tracker for backport

**Comment by trociny** [comment]
@jdurgin I understand luminous is EOL, but I still would like to backport this to luminous branch too (even if it will never be released). Can I update the tracker to add luminous backport?

**Comment by jdurgin** [comment]
@trociny sure, go for it

**Pull request merged**